### ISSUE TITLE
Performance regression

### ISSUE THREAD

**Issue by Gobot1234** [microsoft/pyright#4573]
**Describe the bug**
Performance when type checking is really slow

**To Reproduce**
I was editing some code in https://github.com/Gobot1234/steam.py/blob/main/steam/state.py and noticed that the LSP isn't updating as fast as it normally does leading to situations like
![image](https://user-images.githubusercontent.com/50501825/216403408-3a894694-ee22-4da5-a3df-100b9a216040.png) 
happening for multiple seconds and I go check my activity monitor and see that code is pinned at 100% 
![image](https://user-images.githubusercontent.com/50501825/216404375-9985a023-8e13-46bd-ae1b-77ab2ec6a68e.png)

**Expected behavior**
Not having to restart the LSP to get type hint info.

The full code I added was
```py
    async def handle_chat_update_from_message(
        self,
        chat_group: ChatGroup,
    ) -> None:
        ...

    async def handle_chat_message(self, msg: chat.IncomingChatMessageNotification) -> None:
        try:
            destination = self._chat_groups[ChatGroupID(msg.chat_group_id)]
        except KeyError:
            return log.debug(f"Got a message for a chat we aren't in {msg.chat_group_id}")

        channel = destination._channels[ChatID(msg.chat_id)]
        channel._update(msg)

        if msg.server_message:
            await self.handle_chat_update_from_message(chat_group)
            return

        (message_cls,) = channel._type_args
        message = message_cls(
            proto=msg,
            channel=channel,  # type: ignore  # type checkers aren't able to figure out this is ok
            author=destination._maybe_member(ID64(msg.steamid_sender)),
        )
        channel.last_message = message  # type: ignore  # same as above
        self._messages.append(message)
        self.dispatch("message", message)
```

Is there anything else you need to debug the issue further?

**VS Code extension or command-line**
pylance v2023.2.11 Pre-Release

**Comment by Gobot1234** [comment]
Adding further to this, I and others have recently noticed issues with "class definition X depends on itself" which seem to get fixed when you restart the LSP but I'm not sure what causes that in the slightest but I've also found that very prevalent in the previously linked state.py file so it might be a similar issue.

**Comment by erictraut** [comment]
Thanks for the bug report. Interestingly, I'm able to repro the problem with Pylance v2023.2.11 Pre-Release, but I'm not able to repro it with pyright 1.1.292. I would expect those to be the same. I'll need to dig further. @debonte, let me know if you have any ideas here.

@Gobot1234, if you have a repro case for the other issue you're seeing, please file a separate issue with repro steps.

**Comment by Gobot1234** [comment]
> @Gobot1234, if you have a repro case for the other issue you're seeing, please file a separate issue with repro steps.

No one as far as I know is able to reproduce it reliably, I just mentioned it cause I thought it might be relevant if I ever do come up with a way to reproduce it I'll report it :)

**Comment by debonte** [comment]
Two thoughts from talking about this during scrum:
1. Pylance started our pull from Pyright a day or two early so we could get a head start on addressing some conflicts. But there was at least one problem with updating that pull PR with the latest changes before shipping. So maybe Pylance and Pyright are out of sync in some way.
2. Our new pytest support is included in 2023.2.11. Pytest isn't used in `state.py`, so in theory this feature shouldn't have any impact. But there could be a bug.

I haven't been able to repro the CPU getting pegged. Here's what I did:
1. Add `from steam.chat import ChatGroup`. Should I not be doing this? I see similar perf behavior without it.
2. Paste the code snippet above on top of the existing `handle_chat_message` function definition.

With both Pylance and Pyright, it takes 2 seconds for the diagnostic to appear on the use of the undefined `chat_group` variable in:

```python
        if msg.server_message:
            await self.handle_chat_update_from_message(chat_group)
            return
```

And 2 seconds for the diagnostic squiggly at that range to disappear when I undo the paste.

I don't see the CPU getting pegged at all though, and without that symptom I'm assuming I'm not hitting the same issue you are.

**Comment by erictraut** [comment]
> With both Pylance and Pyright, it takes 2 seconds for the diagnostic to appear

Hmm, I'm not seeing this with pyright. It completes almost immediately for me. Definitely not a multi-second delay. Which version of pyright are you running? The latest published? Built from main? Do you have `useLibraryCodeForTypes` enabled? (It's off by default in pyright, so it needs to be enabled to replicate the defaults in pylance.)

**Comment by Gobot1234** [comment]
This is the kind of performance I'm getting @debonte 
https://user-images.githubusercontent.com/50501825/216433215-4a89e17f-c287-4392-ba7a-fcf974fd33d6.mov
https://user-images.githubusercontent.com/50501825/216433540-febc3a8f-881f-421d-a7b0-7dc1f44bab27.mov

**Comment by erictraut** [comment]
@Gobot1234, could you please try this experiment? Temporarily disable pylance and try pyright instead.

1. Change the `python.languageServer` setting from `Default` or `Pylance` to `None`.
2. Uninstall or disable the Pylance extension
3. Install the latest Pyright extension from the marketplace (1.1.292)

See if you can repro the same slow behavior.

To undo:
4. Uninstall the Pyright extension
5. Set `python.languageServer` back to `Default` or `Pylance`. This will prompt you to reinstall or enable Pylance.

**Comment by Gobot1234** [comment]
Yeah I can't reproduce this with Pyright, the performance is what I'm used to from Pylance.

**Comment by erictraut** [comment]
OK, thanks for the confirmation. That's a useful clue.

**Comment by debonte** [comment]
> Which version of pyright are you running? The latest published? Built from main?

Latest published -- 1.1.292

> Do you have `useLibraryCodeForTypes` enabled?

I didn't have it set initially. Setting it to true did not change pyright's 2 second delay for me.

> This is the kind of performance I'm getting @debonte

Ok, I'm not seeing anything that bad. @erictraut, is this what you're seeing also?

Not sure why this would be OS-specific, but I think you are both on Macs? I'm on Windows.

**Comment by erictraut** [comment]
I just found one bug that I discovered when exploring this issue. I noticed that the "unaccessed import" diagnostic for the `PurchaseResult` symbol was dependent on evaluation ordering. If I hovered over that import when modifying the source file, the diagnostic went away, which is obviously incorrect. I don't see any way this could impact performance though.

@Gobot1234, I just noticed that you're making use of TypeVar defaults (PEP 696) in this code base. Nice!

@debonte, yes, I'm seeing pauses of 2 to 5 seconds sometimes with the pre-release version of Pylance. The problem occurs only shortly after I open the file (within the first 25 seconds or so), and then it seems to go away.

I don't think it's directly related to type analysis. I've enabled trace logging, and the times for analysis are not very long. Yet, there's something that's clearly taking a long time. Perhaps it's the new pytest support? @rchiodo, does the pytest code perform any synchronous activity on the foreground thread? If so, does it honor cancellations?

Update... I think I just found the smoking gun... I was able to repro the problem in the debugger with the latest pylance sources. I broke into the debugger during one of these long pauses, and it is executing code related to pytest fixtures. @rchidio, looks like you have some debugging to do. Here's a screen shot of the stack trace.

![Screenshot 2023-02-02 at 1 06 24 PM](https://user-images.githubusercontent.com/7040122/216449433-b559cab1-b5e2-443a-a907-32e211b658a2.png)

**Comment by erictraut** [comment]
Transferring the bug to pylance-release, since this isn't in the pyright code.
